src/analysis/souffle/tests/test_taint_with_auth_logic.dl

```diff
@@ -0,0 +1,101 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "../taint_with_auth_logic.dl"
+#include "../flat_graph_ir.dl"
+#include "../edges_from_graph.dl"
+
+// This is a datalog file that is meant to test taint_with_auth_logic.dl. It 
+// should eventually be replaced by or otherwise used as a part of an
+// automated test.
+
+//-----------------------------------------------------------------------------
+// Program Manifests
+//-----------------------------------------------------------------------------
+
+// Module 1
+//      reads: 
+//          some_struct{ a, b }
+//          other_input
+//      writes:
+//          mod1out
+bindFlatParticle("module1", $FlatParticle(
+    /* reads */
+    $Cons("some_struct.a", $Cons("some_struct.b",
+        $Cons("other_input", $Nil()))),
+    /* writes */
+    $Cons("mod1out", $Nil())
+)).
+
+// the label of "other_input" is the empty set (which makes it public).
+// so it has no label elements
+
+// Module 2
+//     reads:
+//         mod1out
+//         input4
+//     writes:
+//         mod2out 
+bindFlatParticle("module2", $FlatParticle(
+    /* reads */
+    $Cons("mod1out", $Cons("input4", $Nil())),
+    /* writes */
+    $Cons("mod2out", $Nil())
+)).
+
+//-----------------------------------------------------------------------------
+// Policy
+//-----------------------------------------------------------------------------
+
+// Assign owners to tags
+ownsTag("Alice", "raw_video").
+ownsTag("Alice", "image_selection").
+ownsTag("Alice", "product_id").
+ownsTag("ServiceProvider", "model_weights").
+
+// Assign tags to initial inputs
+claimHasTag("some_struct.a", "raw_video").
+claimHasTag("some_struct.b", "image_selection").
+claimHasTag("input4", "model_weights").
+
+// The label of mod1out is downgraded to public (= empty set) 
+// because both the raw_video and image_selection tags are removed, and then
+// upgraded to {"product_id"} with this line:
+claimHasTag("mod1out", "product_id").
+
+// Set the downgrades permitted by both parties
+saysDowngrades("Alice", "mod1out", "raw_video").
```

Comment:
How would these `saysDowngrades` be expressed in the IR/manifest?

Reply:
I think they would not be expressed in the IR/manifest exactly. My understanding is essentially that we have two different languages that both compile to datalog, and we combine the resulting datalog to do our analysis/check:
1) The IR/manifest describes the application graph and comes from the application framework that is usually used for that application domain. It's written in protobufs.
2) The authorization logic code that is used to express policy (aside from ingress/egress labels). This is written in a small standalone compiled language.

The `saysDowngrades` thing is the output of the authorizaiton logic compiler for `Alice says downgrades(mod1out, raw_video)`.  We can just import the tuples from the authorizaiton logic compiler into this analysis.

The ingress/egress labels (which are specified with `claimHasTag`) could plausibly be a part of the manifest/IR, or we could plausibly have another very small "language" for specifying these that is kept separately just so that we could include it in remote attestation measurements.

I guess we could also try to capture all aspects of the policy aside from the graph and this check with authorization logic. It seems plausibly that `claimHasTag` could be given a speaker (but we would need to try this out with examples :) ). I think there is no way around making tag ownership (with `ownsTag`) globally trusted, though (and I think we could write axioms that give this behavior).

Reply:
Thanks for the explanation. It makes sense.